compiler/rustc_codegen_ssa/src/mir/debuginfo.rs

```diff
@@ -301,7 +293,40 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                 }
             }
 
-            bx.dbg_var_addr(dbg_var, dbg_loc, base.llval, direct_offset, &indirect_offsets);
+            // When targeting MSVC, spill extra locals for arguments instead of pointing multiple
+            // dbg_var_addr() calls into the same alloca with offsets. MSVC uses CodeView records
+            // not DWARF and LLVM doesn't support translating the resulting
+            // [DW_OP_deref, DW_OP_plus_uconst, offset, DW_OP_deref] debug info to CodeView.
+            // Spilling extra values to the stack makes the resulting debug info simple enough
+            // that LLVM can generate correct CodeView records and thus the values appear in the
+            // debugger. (#83709)
+            let should_spill_extra = bx.cx().sess().target.is_like_msvc
+                && self.mir.local_kind(local) == mir::LocalKind::Arg
+                // LLVM can handle simple things but anything more complex than just a direct
+                // offset or one indirect offset of 0 is too complex for it to generate CV records
+                // correctly.
+                && (direct_offset != Size::ZERO
+                    || indirect_offsets.len() > 1
+                    || (indirect_offsets.len() == 1 && indirect_offsets[0] != Size::ZERO));
+
+            if should_spill_extra {
+                // Create a spill variable which will be a pointer to the actual value
+                let ptr_ty = bx.tcx().mk_ty(ty::RawPtr(ty::TypeAndMut {
+                    mutbl: mir::Mutability::Mut,
+                    ty: place.layout.ty,
+                }));
+                let ptr_layout = bx.layout_of(ptr_ty);
+                let spill = PlaceRef::alloca(bx, ptr_layout);
+                bx.set_var_name(spill.llval, &(var.name.to_string() + ".dbg.spill"));
+
+                // Write the pointer to the spilled variable
+                bx.store(place.llval, spill.llval, spill.align);
+
+                // Point the debug info to `*spill` for the current variable
+                bx.dbg_var_addr(dbg_var, dbg_loc, spill.llval, Size::ZERO, &[Size::ZERO]);
```

Comment:
It is unfortunate that this is necessary, but I guess gotta do what we gotta do. Did you check if clang has a similar problem/needs to implement a similar workaround when targeting msvc?

Reply:
Yeah clang has the same issue when targeting msvc.